src/crypto/PSAKeyAllocator.h

```diff
@@ -0,0 +1,216 @@
+     * In some cases the key attributes must be redefined to match the specific requirements of the
+     * secure location and the cryptographic hardware. This method allows the platform to update the
+     * key attributes before storing the key.
+     *
+     * Read the current key attributes to determine the key type, algorithm, and usage flags. Update
+     * the key attributes as needed.
+     *
+     * @param attrs Reference to the key attributes structure to be updated.
+     */
+    virtual void UpdateKeyAttributes(psa_key_attributes_t & attrs) = 0;
+
+    // Allow setting and getting the static instance of the PSAKeyAllocator by external functions
+    friend PSAKeyAllocator & GetPSAKeyAllocator();
+    friend void SetPSAKeyAllocator(PSAKeyAllocator * keyAllocator);
+
+private:
+    static PSAKeyAllocator * sInstance;
+};
+
+/**
+ * @brief Default implementation of PSAKeyAllocator.
+ *
+ * This default implementation allocates key IDs according to the KeyIdBase enum.
+ * The operational key ID is calculated as the base operational key ID plus the fabric index.
+ * The DAC key ID is calculated as the base DAC key ID.
+ * The ICD key ID is allocated from the range starting from the ICDKeyRangeStart.
+ * The key attributes are not updated.
+ */
+class DefaultPSAKeyAllocator : public PSAKeyAllocator
+{
+public:
+    // implementations of the PSAKeyAllocator interface
+    psa_key_id_t GetDacKeyId() override { return to_underlying(KeyIdBase::DACPrivKey); }
+    psa_key_id_t GetOpKeyId(FabricIndex fabricIndex) override { return MakeOperationalKeyId(fabricIndex); }
+    psa_key_id_t AllocateICDKeyId() override
+    {
+        psa_key_id_t newKeyId = PSA_KEY_ID_NULL;
+        if (CHIP_NO_ERROR !=
+            Crypto::FindFreeKeySlotInRange(newKeyId, to_underlying(KeyIdBase::ICDKeyRangeStart), kMaxICDClientKeys))
+        {
+            newKeyId = PSA_KEY_ID_NULL;
+        }
+        return newKeyId;
+    }
+    void UpdateKeyAttributes(psa_key_attributes_t & attrs) override
+    {
+        // Do nothing
+    }
+};
+
+/**
+ * @brief Static function to get the instance of PSAKeyAllocator.
+ *
+ * If the static instance is not set, the default implementation is returned.
+ *
+ * @return PSAKeyAllocator reference to the instance of PSAKeyAllocator.
+ */
+inline PSAKeyAllocator & GetPSAKeyAllocator()
+{
+    if (!PSAKeyAllocator::sInstance)
+    {
+        static DefaultPSAKeyAllocator defaultAllocator;
+        return defaultAllocator;
+    }
+    return *PSAKeyAllocator::sInstance;
+}
+
+/**
+ * @brief Set the static implementation of the PSAKeyAllocator.
+ *
+ * Providing nullptr as an argument will revert to the default implementation.
+ *
+ * @param keyAllocator Pointer to the PSAKeyAllocator instance to be set.
+ */
+inline void SetPSAKeyAllocator(PSAKeyAllocator * keyAllocator)
+{
+    PSAKeyAllocator::sInstance = keyAllocator;
+}
+
+// Initialize the PSAKeyAllocator instance
+inline PSAKeyAllocator * PSAKeyAllocator::sInstance = nullptr;
```

Comment:
Would it make sense to create a PSAKeyAllocator.cpp and use `inline` less? I am not sure how things will interact if this gets included in more than one file.

Reply:
Thanks to C++ 17 (https://en.cppreference.com/w/cpp/language/inline) we can use inline here and avoid creating and maintaining the additional PSAKeyAllocator.cpp file. Inline variables have been tested on most compilers https://en.cppreference.com/w/cpp/compiler_support/17#cpp_inline_variables_201606L and I assume here that it works as it is designed and this instance will be created only once within the Matter translation unit no matter how many times the header is included. Please correct me if I'm wrong with my thinking, maybe I have wrong assumptions regarding C++17 here, or somewhere in Matter configuration there is something that can cause problems with that, and maybe I'm unaware of it.

Reply:
@andy31415 What do you think?

Reply:
My concern is that this feels like "tricky code". If I am required to read and understand the C++ standard in detail (including differences across versions) to ensure this is safe then the code may not be as "obviously correct" as one would like.

Otherwise sure, I am ok with this as long as you explain in a comment why this is safe and also explain why if this does not work as expected we would blow up/have a test that fails. Right now I am wondering if this works only because a single cpp files includes it or not.

Reply:
I found that there is no test that covers your concerns. I've created a new one - `TestKeyAllocation` under the `TestPSAOpKeyStore` suite. It checks the whole functionality of the PSAKeyAllocator and the DefaultPSAKeyAllocator as well. This suite is run within the nrfconnect POSIX unit tests workflow. I've removed locally `inline` from there and the build failed as expected due to multiple definitions. It proves that inline works there and now we have at least one test where it should fail if it does not work as expected. I've added the comment there as well to explain my motivation. Of course, this is all correct if we use C++17, but this is my assumption.

@andy31415 Does this clear up your doubts? :)